homeassistant/components/teslemetry/sensor.py

```diff
@@ -0,0 +1,443 @@
+        native_unit_of_measurement=UnitOfEnergy.WATT_HOUR,
+        suggested_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.ENERGY_STORAGE,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        icon="mdi:battery-high",
+    ),
+    SensorEntityDescription(
+        key="percentage_charged",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.BATTERY,
+        suggested_display_precision=2,
+    ),
+    SensorEntityDescription(
+        key="battery_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:home-battery",
+    ),
+    SensorEntityDescription(
+        key="load_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:power-plug",
+    ),
+    SensorEntityDescription(
+        key="grid_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:transmission-tower",
+    ),
+    SensorEntityDescription(
+        key="grid_services_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:transmission-tower",
+    ),
+    SensorEntityDescription(
+        key="generator_power",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfPower.WATT,
+        suggested_unit_of_measurement=UnitOfPower.KILO_WATT,
+        suggested_display_precision=2,
+        device_class=SensorDeviceClass.POWER,
+        icon="mdi:generator-stationary",
+    ),
+)
+
+WALL_CONNECTOR_DESCRIPTIONS: tuple[SensorEntityDescription, ...] = (
+    SensorEntityDescription(
+        key="wall_connector_state",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        device_class=SensorDeviceClass.ENUM,
+        icon="mdi:ev-station",
+    ),
+    SensorEntityDescription(
+        key="wall_connector_fault_state",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        entity_registry_enabled_default=False,
+        device_class=SensorDeviceClass.ENUM,
+        icon="mdi:ev-station",
+    ),
```

Comment:
Do we have a mapping for these states? Having a code for state is not really user friendly.
I found a mapping for the tesla wall connector integration. I'm not sure if it's the same mapping using the fleet api. https://github.com/home-assistant/core/pull/107909

Reply:
Do you have a wall connector? I can test your PR with my wall connector to see what I have for fault state. May be there is different type of errors so that's why it's an enum 🤷‍♂️

Reply:
Yes I have a wall connector.

I set it as an Enum because I know it will have a set of numerical codes that probably relate to the errors in the manual, but I have no reference about what's what and am not sure I could force those errors.

This is why I disabled these entities by default, since they will be useless for most users.

Except now that you have states I'll probably enable the first one.

Reply:
I just tested with my wall connector. State codes are not the same between fleet api and local api... I could figured what's fault state because it changed with the vehicle is charging.

**Vehicle unplugged:**
Teslemetry state = **2**
Teslemetry fault state = **2**
Wall connector state = **1**

**Vehicle plugged:**
Teslemetry state = **4**
Teslemetry fault state = **2**
Wall connector state = **9**

**Vehicle charging :**
Teslemetry state = **1**
Teslemetry fault state = **8**
Wall connector state = **11**

Reply:
Ive added "code" to the translations